### DISCUSSION TITLE
Preinstall proprietary software in full image.

### DISCUSSION THREAD

**Discussion by Deepak-Beniya** [OE4T/tegra-demo-distro#297] [Q&A]
Hlo,
I have a proprietary software, it should be installed in the full image . 
Same as like calculator, web browser applications which comes preinstalled  in the image .
I want my software(ebus.deb) to be preinstalled in the full image.
I have created an meta layer , inside the meta layer I have created a receipe. Inside the receipe I have written my bb file.
I am facing issue in do_install. 

I need help to install this deb file .

here is the bb file

######################BB file start##############################################

#Recipe for eBUS Runtime Jetson

DESCRIPTION = "Installs eBUS Runtime for Jetson"

LICENSE = "CLOSED"

DEPENDS += "dpkg-native"

TASK_PACKAGE = "package"

SRC_URI = "file://eBUS_Runtime_Jetson_5.1_linux-aarch64-arm-6.4.0-6670.deb"
SSTATE_CACHE = "${SSTATE_DIR}/eBUS_runtime"

inherit package

PACKAGE_ARCH = "${MACHINE_ARCH}"

S = "${WORKDIR}"

FILES_${PN} = "${bindir}/*"

INSANE_SKIP_${PN} = "ldflags"

do_install () {
  dpkg -i "${WORKDIR}/eBUS_Runtime_Jetson_5.1_linux-aarch64-arm-6.4.0-6670.deb"
  #Optional: Uncomment and adjust these lines if the DEB installs files elsewhere
   install -d ${D}${DESTDIR}/opt/ebus
   mv ${D}${WORKDIR}/usr/local/sbin ${D}${DESTDIR}/opt/ebus
}

#Post-installation tasks (optional)
pkg_postinst_ontarget:${PN} () {
  #Example: Create a directory and set permissions (adjust paths/permissions)
  install -d /var/log/ebus
  chmod 0644 /opt/ebus/config.ini

  #Example: Run a post-installation script from your DEB (adjust path)
  if [ -x /opt/ebus/postinst ]; then
    /opt/ebus/postinst
  fi
}

############################BB file end####################################

ERROR is 

ERROR: pelora-1.0.0-r0 do_install: ExecutionError('/home/deepak/fiverr/client2/tegra-demo-distro/build_tegra_demo/tmp/work/jetson_orin_nano_devkit-oe4t-linux/pelora/1.0.0-r0/temp/run.do_install.16713', 2, None, None)
ERROR: Logfile of failure stored in: /home/deepak/fiverr/client2/tegra-demo-distro/build_tegra_demo/tmp/work/jetson_orin_nano_devkit-oe4t-linux/pelora/1.0.0-r0/temp/log.do_install.16713

 DEBUG: Python function extend_recipe_sysroot finished
| DEBUG: Executing shell function do_install
| dpkg: warning: 'ldconfig' not found in PATH or not executable
| dpkg: error: 1 expected program not found in PATH or not executable
| Note: root's PATH should usually contain /usr/local/sbin, /usr/sbin and /sbin
| WARNING: exit code 2 from a shell command.
ERROR: Task (/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta-pelora/recipes-pelora/pelora/pelora_1.0.0.bb:do_install) failed with exit code '1'
NOTE: Tasks Summary: Attempted 668 tasks of which 667 didn't need to be rerun and 1 failed.

Summary: 1 task failed:
  /home/deepak/fiverr/client2/tegra-demo-distro/layers/meta-pelora/recipes-pelora/pelora/pelora_1.0.0.bb:do_install
Summary: There was 1 ERROR message, returning a non-zero exit code.

**Comment by Deepak-Beniya** [comment]
Right now I am facing issue on pre-dependency

errors are






ERROR: pelora-1.0.0-r0 do_install: ExecutionError('/home/deepak/fiverr/client2/tegra-demo-distro/build_tegra_demo/tmp/work/jetson_orin_nano_devkit-oe4t-linux/pelora/1.0.0-r0/temp/run.do_install.8098', 1, None, None)
ERROR: Logfile of failure stored in: /home/deepak/fiverr/client2/tegra-demo-distro/build_tegra_demo/tmp/work/jetson_orin_nano_devkit-oe4t-linux/pelora/1.0.0-r0/temp/log.do_install.8098
Log data follows:
| DEBUG: Executing python function extend_recipe_sysroot
| NOTE: Direct dependencies are ['/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta/recipes-core/glibc/glibc_2.35.bb:do_populate_sysroot', '/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta/recipes-core/glibc/ldconfig-native_2.12.1.bb:do_populate_sysroot', '/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta/recipes-devtools/gcc/gcc-cross_11.4.bb:do_populate_sysroot', '/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta/recipes-devtools/gcc/gcc-runtime_11.4.bb:do_populate_sysroot', '/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta/recipes-devtools/quilt/quilt-native_0.67.bb:do_populate_sysroot', 'virtual:native:/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta/recipes-devtools/dpkg/dpkg_1.21.4.bb:do_populate_sysroot', 'virtual:native:/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta/recipes-devtools/patch/patch_2.7.6.bb:do_populate_sysroot', 'virtual:native:/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta/recipes-devtools/pseudo/pseudo_git.bb:do_populate_sysroot', 'virtual:native:/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta/recipes-extended/xz/xz_5.2.6.bb:do_populate_sysroot']
| NOTE: Installed into sysroot: []
| NOTE: Skipping as already exists in sysroot: ['glibc', 'ldconfig-native', 'gcc-cross-aarch64', 'gcc-runtime', 'quilt-native', 'dpkg-native', 'patch-native', 'pseudo-native', 'xz-native', 'gnu-config-native', 'mpfr-native', 'texinfo-dummy-native', 'linux-libc-headers', 'libtool-native', 'libmpc-native', 'zlib-native', 'gmp-native', 'flex-native', 'zstd-native', 'binutils-cross-aarch64', 'gettext-minimal-native', 'opkg-utils-native', 'bzip2-native', 'perl-native', 'gettext-native', 'libgcc', 'attr-native', 'm4-native', 'make-native', 'gdbm-native']
| DEBUG: Python function extend_recipe_sysroot finished
| DEBUG: Executing shell function do_install
| dpkg: regarding .../eBUS_Runtime_Jetson_5.1_linux-aarch64-arm-6.4.0-6670.deb containing ebus-runtime, pre-dependency problem:
|  ebus-runtime pre-depends on debconf (>= 0.2.17)
|   debconf is not installed.
| 
| dpkg: error processing archive /home/deepak/fiverr/client2/tegra-demo-distro/build_tegra_demo/tmp/work/jetson_orin_nano_devkit-oe4t-linux/pelora/1.0.0-r0/eBUS_Runtime_Jetson_5.1_linux-aarch64-arm-6.4.0-6670.deb (--install):
|  pre-dependency problem - not installing ebus-runtime
| Errors were encountered while processing:
|  /home/deepak/fiverr/client2/tegra-demo-distro/build_tegra_demo/tmp/work/jetson_orin_nano_devkit-oe4t-linux/pelora/1.0.0-r0/eBUS_Runtime_Jetson_5.1_linux-aarch64-arm-6.4.0-6670.deb
| WARNING: exit code 1 from a shell command.
ERROR: Task (/home/deepak/fiverr/client2/tegra-demo-distro/layers/meta-pelora/recipes-pelora/pelora/pelora_1.0.0.bb:do_install) failed with exit code '1'
NOTE: Tasks Summary: Attempted 677 tasks of which 676 didn't need to be rerun and 1 failed.

Summary: 1 task failed:
  /home/deepak/fiverr/client2/tegra-demo-distro/layers/meta-pelora/recipes-pelora/pelora/pelora_1.0.0.bb:do_install
Summary: There was 1 ERROR message, returning a non-zero exit code.

**Reply by dwalkes** [reply]
This probably isn't the best forum for this question as it's not really specific or related to tegra but rather it's a general question about including a deb package in your bitbake build.  The suggestions on the thread at https://stackoverflow.com/a/36090585 might be helpful, and as mentioned there if possible it would be better to build from source rather than install from deb package I suppose.

**Reply by Deepak-Beniya** [reply]
I have checked everything on internet , i haven't got any solution .
The software is propriatory software, I don't have source code. I have purchased the software and I need to install in full image .

If you need the software i can send it to you .You can create a recepie and try it in your source.

**Comment by Deepak-Beniya** [comment]
Is my recipe correct to install the software in rootfs ?

**Reply by madisongh** [reply]
Not really, no. You can't just use Debian tools to install a package during a Yocto build, as there's nothing in the build system that provides the support that those tools expect.

Instead, you need to use tools like `dpkg-deb` to analyze the contents of the package, including the Debian control files (like the postinst script) and replicate that logic, if it's applicable, in your recipe manually.  You'll find examples of how we do this with the NVIDIA deb packages from JetPack throughout the meta-tegra layer.

**Reply by Deepak-Beniya** [reply]
okay,then tell me how can I simply copy the .deb file into rootfs? 
If it's available in rootfs then I can manually install it.
with the help this receipe.

**Reply by madisongh** [reply]
That probably won't work very well, either, since your root filesystem will not look like one generated by the Debian/Ubuntu tools.
